src/main/io/beeper.c

```diff
@@ -418,37 +420,59 @@ void beeperUpdate(timeUs_t currentTimeUs)
-        schedulerIgnoreTaskExecTime();
-        return;
+    bool dshotBeaconRequested = false;
+
+    if (!areMotorsRunning()) {
+        const beeperMode_e activeMode = currentBeeperEntry ? currentBeeperEntry->mode : BEEPER_SILENCE;
+        const bool usbIn = usbCableIsInserted();
+
+        // Drive the ESC beacon whenever the beeper has entered the RX_LOST sequence.
+        if (activeMode == BEEPER_RX_LOST
+            && !usbIn
+            && !(beeperConfig()->dshotBeaconOffFlags & BEEPER_GET_FLAG(BEEPER_RX_LOST)) ) {
+            dshotBeaconRequested = true;
+        }
+
+        // Allow user-triggered beacon via AUX switch while the RX link is healthy.
+        if (IS_RC_MODE_ACTIVE(BOXBEEPERON)
+            && failsafeIsReceivingRxData()
+            && !(beeperConfig()->dshotBeaconOffFlags & BEEPER_GET_FLAG(BEEPER_RX_SET)) ) {
+            dshotBeaconRequested = true;
+        }
     }
 
-#ifdef USE_DSHOT
-    if (!areMotorsRunning()
-        && (DSHOT_BEACON_ALLOWED_MODES & BEEPER_GET_FLAG(currentBeeperEntry->mode))
-        && !(beeperConfig()->dshotBeaconOffFlags & BEEPER_GET_FLAG(currentBeeperEntry->mode)) ) {
-        const timeDelta_t dShotBeaconInterval = (currentBeeperEntry->mode == BEEPER_RX_SET)
-            ? DSHOT_BEACON_MODE_INTERVAL_US
-            : DSHOT_BEACON_RXLOSS_INTERVAL_US;
+    if (dshotBeaconRequested) {
         if (cmpTimeUs(currentTimeUs, getLastDisarmTimeUs()) > DSHOT_BEACON_GUARD_DELAY_US
-            && !isTryingToArm() ) {
-            if (cmpTimeUs(currentTimeUs, lastDshotBeaconCommandTimeUs) > dShotBeaconInterval) {
-                // at least 500ms between DShot beacons to allow time for the sound to fully complete
+            && !isTryingToArm()) {
+            if (cmpTimeUs(currentTimeUs, lastDshotBeaconCommandTimeUs) > dshotBeaconIntervalUs) {
+                // at least 450ms between DShot beacons to allow time for the sound to fully complete
                 // the DShot Beacon tone duration is determined by the ESC, and should not exceed 250ms
                 lastDshotBeaconCommandTimeUs = currentTimeUs;
                 dshotCommandWrite(ALL_MOTORS, getMotorCount(), beeperConfig()->dshotBeaconTone, DSHOT_CMD_TYPE_INLINE);
             }
         } else {
             // make sure lastDshotBeaconCommandTimeUs is valid when DSHOT_BEACON_GUARD_DELAY_US elapses
-            lastDshotBeaconCommandTimeUs = currentTimeUs - dShotBeaconInterval;
+            lastDshotBeaconCommandTimeUs = currentTimeUs - dshotBeaconIntervalUs;
         }
     }
 #endif
+    // Note: DShot beacon handling above must run even if no beeper sequence is active.
+    // Beeper routine doesn't need to update if there aren't any sounds ongoing
+    if (currentBeeperEntry == NULL) {
+        schedulerIgnoreTaskExecTime();
+        return;
+    }
+
+    if (beeperNextToggleTime && cmp32(beeperNextToggleTime, currentTimeUs) > 0) {
+        schedulerIgnoreTaskExecTime();
+        return;
+    }
```

Comment:
I think it is too late to run these two functions here. It could potentially give the scheduler a wrong impression about task runtime. Maybe I am looking at it the wrong way... @SteveCEvans ?